File: content/problems/2925-maximum-score-after-applying-operations-on-a-tree/solutions/iterative-tree-dp-minimum-kept-value/Solution.java
class Solution {
    public long maximumScoreAfterOperations(int[][] edges,int[] values){
        int n=values.length;
        List<Integer>[] graph=new ArrayList[n];
        for(int i=0;i<n;i++)graph[i]=new ArrayList<>();
        for(int[] edge:edges){
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }
        int[] parent=new int[n],order=new int[n];
        Arrays.fill(parent,-1);
        parent[0]=0;
        int size=0;
        ArrayDeque<Integer> stack=new ArrayDeque<>();
        stack.push(0);
        while(!stack.isEmpty()){
            int node=stack.pop();
            order[size++]=node;
            for(int next:graph[node])if(parent[next]==-1){
                parent[next]=node;
                stack.push(next);
            }
        }
        long total=0;
        for(int value:values)total+=value;
        long[] keep=new long[n];
        for(int index=n-1;index>=0;index--){
            int node=order[index];
            long children=0;
            boolean leaf=true;
            for(int next:graph[node])if(parent[next]==node){
                leaf=false;
                children+=keep[next];
            }
            keep[node]=leaf?values[node]:Math.min((long)values[node],children);
        }
        return total-keep[0];
    }
}

File: content/problems/2925-maximum-score-after-applying-operations-on-a-tree/solutions/iterative-tree-dp-minimum-kept-value/Solution.kt
class Solution {
    fun maximumScoreAfterOperations(edges:Array<IntArray>,values:IntArray):Long{
        val n=values.size;
        val graph=Array(n){
            mutableListOf<Int>()
        };
        for(edge in edges){
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0])
        };
        val parent=IntArray(n){
            -1
        };
        val order=IntArray(n);
        parent[0]=0;
        var size=0;
        val stack=java.util.ArrayDeque<Int>();
        stack.push(0);
        while(stack.isNotEmpty()){
            val node=stack.pop();
            order[size++]=node;
            for(next in graph[node])if(parent[next]==-1){
                parent[next]=node;
                stack.push(next)
            }
        };
        var total=0L;
        for(value in values)total+=value;
        val keep=LongArray(n);
        for(index in n-1 downTo 0){
            val node=order[index];
            var children=0L;
            var leaf=true;
            for(next in graph[node])if(parent[next]==node){
                leaf=false;
                children+=keep[next]
            };
            keep[node]=if(leaf)values[node].toLong() else minOf(values[node].toLong(),children)
        };
        return total-keep[0]
    }
}
